### PR TITLE
Blog: migrations, postgres locks and application downtime

### PR DESCRIPTION
**Pull request by andrew-farries**
## Summary

- Describe two classes of migration errors; incompatible schema changes and long-lived locks.
- Give an example of the latter kind of problem.
- Show how to avoid it by setting `lock_timeout`
- Show how `pgroll` solves the problem using lock timeouts, backoff and retry.
- Make a general point about how migration tools should be smarter; helping users avoid both classes of migration error.

### Blog Release Checklist

- \[ ] Release on: \<mm/dd/yy>
- \[ ] Published on dev.to (connected to Xata org)
- \[ ] OG image included
- \[ ] Social media is scheduled for this post

### CODE REVIEW

File: postgres-migrations-locking-and-backoff.mdx [hunk 1]
```diff
@@ -0,0 +1,166 @@
+---
+title: Schema changes and the Postgres lock queue
+description: 'Learn how schema changes can cause downtime by locking out reads and writes and how migration tools can avoid it by using lock timeouts, along with backoff and retry strategies'
```
Comment (CezzaineZ) [inline postgres-migrations-locking-and-backoff.mdx:3]:
```suggestion
description: 'Learn how schema changes can cause downtime by locking out reads and writes and how migration tools can avoid it by using lock timeouts, along with backoff and retry strategies.'
```

File: postgres-migrations-locking-and-backoff.mdx [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+Schema changes are hard. They're hard because there are multiple different failure modes, some of which require knowledge of how the underlying RDMS system works, especially with regard to object locking.
+
+Broadly speaking, there are two classes of breakage that can occur when applying database migrations:
+
+1. Migrations that make incompatible changes to the schema, breaking client applications.
+2. Migrations that lock a database object for an unacceptable amount of time, causing the application to become unavailable as reads and writes start to fail.
+
+The first kind of breakage is perhaps more widely understood and discussed, but the second kind is just as important to consider when running migrations against production systems.
+
+Today we're going to talk about a specific kind of the second type of breakage: how long running queries together with DDL statements can lock out reads and writes from a table, causing application downtime.
```
Comment (CezzaineZ) [inline postgres-migrations-locking-and-backoff.mdx:25]:
```suggestion
Today we're going to talk about the second type of breakage: how long running queries together with DDL statements can lock out reads and writes from a table, causing application downtime.
```

File: postgres-migrations-locking-and-backoff.mdx [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+```sql
+ALTER TABLE users ADD COLUMN AGE integer
+```
+
+This `ALTER TABLE` statement attempts to acquire an `ACCESS EXCLUSIVE` lock on the `users` table but is unable to do so until the `SELECT` statement completes and releases its `ACCESS SHARE` lock.
+
+So far, so good. What's the problem here? The DDL statement can simply wait patiently until it's able to acquire its `ACCESS EXCLUSIVE` lock, right? The problem is that any other statements that require a lock on the `users` table **are now queued behind this ALTER TABLE statement**, including other `SELECT` statements that only require `ACCESS SHARE` locks.
+
+This means that the table is effectively blocked for reads and writes until the `ALTER TABLE` statement completes. `SELECT`s and `UPDATE`s will queue up behind it, unable to execute. If there is a long-running query that prevents the `ALTER TABLE` from acquiring the lock, than reads and writes will be blocked for the duration of that query.
```
Comment (CezzaineZ) [inline postgres-migrations-locking-and-backoff.mdx:90]:
```suggestion
This means that the table is effectively blocked for reads and writes until the `ALTER TABLE` statement completes. `SELECT`s and `UPDATE`s will queue up behind it, unable to execute. If there is a long-running query that prevents the `ALTER TABLE` from acquiring the lock, then reads and writes will be blocked for the duration of that query.
```

File: postgres-migrations-locking-and-backoff.mdx [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+canceling statement due to lock timeout
+```
+
+Once the `ALTER TABLE` statement has failed, the `SELECT` statement from the third session is unblocked and executes directly.
+
+> DDL statements in migration sessions should always set `lock_timeout` to an appropriate value for the application; values of less than 2 seconds are common. This ensures that reads and writes won't queue behind a blocked DDL statement and cause application downtime.
+
+## Backoff and retry
+
+With `lock_timeout` set appropriately like this, it's possible to ensure that DDL statements like `ALTER TABLE` don't lock out reads and writes from the affected table. However, when a statement exceeds its `lock_timeout` the statement fails. Postgres does not provide any mechanism by which such statements can be automatically retried. Retrying lock acquisition becomes the responsiblity of the the person or process running the DDL statement.
```
Comment (CezzaineZ) [inline postgres-migrations-locking-and-backoff.mdx:148]:
```suggestion
With `lock_timeout` set appropriately like this, it's possible to ensure that DDL statements like `ALTER TABLE` don't lock out reads and writes from the affected table. However, when a statement exceeds its `lock_timeout` the statement fails. Postgres does not provide any mechanism by which such statements can be automatically retried. Retrying lock acquisition becomes the responsiblity of the person or process running the DDL statement.
```

File: postgres-migrations-locking-and-backoff.mdx [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+However you make schema changes to your Postgres database, it's important to consider how long running queries together with DDL statements can block reads and writes to a table. Setting `lock_timeout` on DDL statements is a good first step, but it's also important to consider how to handle lock acquisition failures.
+
+# Conclusion
+
+At the top of the post, we identified two common failure modes for database migrations:
+
+1. Schema changes breaking client applications.
+2. Unintended downtime due to DDL-induced table locks.
+
+[pgroll](https://github.com/xataio/pgroll) is a migration tool for Postgres that attempts to address both of these failure modes. It's 'headline feature' is its ability to make two versions, old and new, of a database schema available to client applications. This mitigates the risk of schema changes breaking applications. We've previously blogged about this [here](https://xata.io/blog/pgroll-internals), and the topic is well-covered in the official `pgroll` [docs](https://github.com/xataio/pgroll/tree/main/docs).
```
Comment (CezzaineZ) [inline postgres-migrations-locking-and-backoff.mdx:164]:
```suggestion
[pgroll](https://github.com/xataio/pgroll) is a migration tool for Postgres that attempts to address both of these failure modes. Its 'headline feature' is its ability to make two versions, old and new, of a database schema available to client applications. This mitigates the risk of schema changes breaking applications. We've previously blogged about this [here](https://xata.io/blog/pgroll-internals), and the topic is well-covered in the official `pgroll` [docs](https://github.com/xataio/pgroll/tree/main/docs).
```

File: postgres-migrations-locking-and-backoff.mdx [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+
+Once the `ALTER TABLE` statement has failed, the `SELECT` statement from the third session is unblocked and executes directly.
+
+> DDL statements in migration sessions should always set `lock_timeout` to an appropriate value for the application; values of less than 2 seconds are common. This ensures that reads and writes won't queue behind a blocked DDL statement and cause application downtime.
+
+## Backoff and retry
+
+With `lock_timeout` set appropriately like this, it's possible to ensure that DDL statements like `ALTER TABLE` don't lock out reads and writes from the affected table. However, when a statement exceeds its `lock_timeout` the statement fails. Postgres does not provide any mechanism by which such statements can be automatically retried. Retrying lock acquisition becomes the responsiblity of the the person or process running the DDL statement.
+
+With `pgroll`, our migration tool for Postgres, such lock acquisition failures are automatically retried with an exponential backoff strategy. This means that:
```
Comment (CezzaineZ) [inline postgres-migrations-locking-and-backoff.mdx:150]:
```suggestion
With `[pgroll](https://github.com/xataio/pgroll)`, our migration tool for Postgres, such lock acquisition failures are automatically retried with an exponential backoff strategy. This means that:
```

Comment (CezzaineZ) [reply postgres-migrations-locking-and-backoff.mdx:150]: First mention of pgroll so wanted to get it in here as well as below.

**Review by CezzaineZ — COMMENTED** [review]
Awesome blog Andrew, thank you! Just spotted a few suggested amends.

**Comment by andrew-farries** [comment]
Thanks for the suggested fixes @CezzaineZ . I've applied all of them.

I re-used the same image as for the `pgroll` internals blog post: [this one](https://raw.githubusercontent.com/xataio/mdx-blog/main/images/pgroll-internal.jpeg). 

Is that OK or should we use a different image?

**Comment by miukimiu** [comment]
@CezzaineZ and @andrew-farries,

I grabbed the previous illustration and just added a lock. @andrew-farries if you think the previous illustration makes more sense, feel free to use it. Or just use this new one:

migrations-and-exclusive-locks

**Comment by andrew-farries** [comment]
Thanks @miukimiu - I've used your new image 🙏

**Review by CezzaineZ — APPROVED** [review]

**Pull request merged**